### PR TITLE
Comparative regression justification update

### PR DESCRIPTION
**Pull request by jadie1**
I've updated the comparative regression justification to include the outcome prediction (if on) and the score reasoning from the best sample. To do this I had to update the alignment functions class to optionally return the index of the best sample.

### CODE REVIEW

File: align_system/utils/alignment_utils.py
Comment (eveenhuis) [inline align_system/utils/alignment_utils.py:99]: This may be me not being capable of math/logic this late in the day, but is this (and alternatives) not doing essentially the same thing as `_select_min_dist_choice` to find the index? Should that function just return an index? I'm also not sure if this logic works for misaligned

I'm also not sure I'm sold on the variable number of returns, but I feel like @dmjoy would have thoughts on that

Comment (jadie1) [reply align_system/utils/alignment_utils.py:99]: It is super similar but different. At first we are trying to select the **choice** with the min distance by aggregating over the samples. Once the choice is selected, we have to find which **sample** had the min distance for that choice because that is the "best" sample and we want to use it's reasoning as justification. Does that make sense? I don't love all the almost identical looking but different logic in the code, but I'm not sure how else to do it.

Comment (eveenhuis) [reply align_system/utils/alignment_utils.py:99]: Ah yes, gotcha

Comment (jadie1) [reply align_system/utils/alignment_utils.py:99]: I don't have strong feelings about the variable number of returns. If that is something we typically avoid we can make it always return the best_sample_index. I just made it an option because it doesn't make much sense in the context of the oracle which doesn't have samples so will always return 0

Comment (jadie1) [reply align_system/utils/alignment_utils.py:99]: I don't think this will work for misaligned... Maybe there is a way to reuse the _select_min_dist logic. I'll work on it

Comment (eveenhuis) [reply align_system/utils/alignment_utils.py:99]: If we know what the distance of the selected choice was, we just need to choose the one closest to *that* if that makes sense

Comment (jadie1) [reply align_system/utils/alignment_utils.py]: If I'm following, you're suggesting selecting the sample with distance closest to the distance of the choice (aggregate over samples), which would give us sort of an average sample. I think we do want the best sample here - or closest to the target

**Comment by jadie1** [comment]
@eveenhuis I switched the best sample selection to also use `_select_min_dist_choice()` so that the misaligned option works

**Comment by eveenhuis** [comment]
Thanks! I will approve and leave the return value decision up to @dmjoy

**Review by eveenhuis — APPROVED** [review]

File: align_system/utils/alignment_utils.py
Comment (jadie1) [reply align_system/utils/alignment_utils.py]: I think that would work, I'll try it

Comment (jadie1) [reply align_system/utils/alignment_utils.py]: I've moved it, I think it is cleaner as a separate function. Let me know what you think https://github.com/ITM-Kitware/align-system/pull/98/commits/9f5b2acc462e64cbfc3e4e90b8a6c77541d837b3

File: align_system/utils/alignment_utils.py
Comment (dmjoy) [inline align_system/utils/alignment_utils.py:90]: is the `kde_norm` argument being used here?

File: align_system/utils/alignment_utils.py
Comment (dmjoy) [inline align_system/utils/alignment_utils.py:128]: I think lowercase underscore separated is more standard for variables (I can maybe understand using it for `AlignmentFunc = AvgDistScalarAlignment` but since you're instantiating `AvgDistScalarAlignment` with `()`, it's an object now not a class)

**Review by dmjoy — APPROVED** [review]
I think this looks a bit better overall, but I think there's still probably some refactoring that could be done to make this code easier to follow and a bit more modular.  That said, I'm not trying to be too picky ahead of the deadline here so I'm content to merge this as is (unless some of the suggestions are quick fixes) and we can work on the refactors / improvements beyond the delivery

File: align_system/utils/alignment_utils.py
Comment (jadie1) [reply align_system/utils/alignment_utils.py]: Fixed here https://github.com/ITM-Kitware/align-system/pull/98/commits/e88a7524b644ebdadb0448ef5aa4fadf7d67a8c3

File: align_system/utils/alignment_utils.py
Comment (jadie1) [reply align_system/utils/alignment_utils.py]: It is not, fixed here https://github.com/ITM-Kitware/align-system/pull/98/commits/f063db1e7029ec986df50496491e0b6902854656

**Pull request merged**
